backend/app/api/ws/device_namespace.py

```
@@ -1169,6 +1219,64 @@ async def _broadcast_device_slot_update(self, user_id: int, device_id: str) -> N
+                "device:upgrade",
+                params,
+                room=socket_id,
+            )
+            logger.info(f"[Device WS] Sent upgrade command to socket {socket_id}")
+            return True
+        except Exception as e:
+            logger.error(f"[Device WS] Failed to send upgrade command: {e}")
+            return False
+
+    async def _broadcast_device_upgrade_status(
+        self, user_id: int, payload: "DeviceUpgradeStatusEvent"
```

Ruff (0.15.6)

[error] Undefined name `DeviceUpgradeStatusEvent`

(F821)